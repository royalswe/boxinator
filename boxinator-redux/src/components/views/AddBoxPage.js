/**
 * TODO:: Form and POST validation after backend validation is done
 * exports AddBoxPage class only for testing purpose
 */

import React from 'react';
import classnames from 'classnames';
import { connect } from 'react-redux';
import { createItem } from '../../actions/actions';

export class AddBoxPage extends React.Component {

    state = {
        name: '',
        color: '',
        weight: '',
        country: '',
        errors: {}
    };

    handleChange = (e) => {
        /**
         * If there is any validation errors,
         * validation errors disappears when isvalid
         */
        if (!!this.state.errors[e.target.name]) {
          let errors = Object.assign({}, this.state.errors);
          delete errors[e.target.name];
          this.setState({
            [e.target.name]: e.target.value,
            errors
          });
        } else {
          this.setState({ [e.target.name]: e.target.value });
        }
      }

    handleSubmit = (e) => {
        e.preventDefault();

        /**
         * Validate form and prevent submit if is not valid
         */
        let errors = {};
        if(this.state.name === '') errors.name = "Name can't be empty";

        this.setState({errors});
        const isValid = Object.keys(errors).length === 0;
        if(isValid){ // if form dont have errors
            const { name, color, weight, country } = this.state;
            this.props.createItem({name, color, weight, country})
        }
    }

    render() {
        return (
            <form onSubmit={this.handleSubmit}>
                <h2>Add new box</h2>

                <div className={classnames('field', { error: !!this.state.errors.name})}>
                    <label htmlFor="name">Name</label>
                    <input
                        name="name"
                        value={this.state.name}
                        onChange={this.handleChange}
                        id="name"
                    />
                </div>
                <span>{this.state.errors.name}</span>
                <div>
                    <label htmlFor="weight">Weight</label>
                    <input
                        name="weight"
                        value={this.state.weight}
                        onChange={this.handleChange}
                        id="weight"
                    />
                </div>
                <div>
                    <label htmlFor="color">Color</label>
                    <input
                        name="color"
                        value={this.state.color}
                        onChange={this.handleChange}
                        id="color"
                    />
                </div>
                <div>
                    <label htmlFor="country">Country</label>
                    <input
                        name="country"
                        value={this.state.country}
                        onChange={this.handleChange}
                        id="country"
                    />
                </div>
                <div>
                    <button>Save</button>
                </div>
            </form>
        )
    }
}

export default connect(null, { createItem })(AddBoxPage);