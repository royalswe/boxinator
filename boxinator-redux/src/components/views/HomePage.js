import React from 'react';

class HomePage extends React.Component {

    render() {
        return (
            <h2>Welcome to Boxinator!</h2>
        )
    }
}

export default HomePage;