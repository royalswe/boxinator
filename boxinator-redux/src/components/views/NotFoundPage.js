import React from 'react';

class HomePage extends React.Component {

    render() {
        // Just a temporary 404 page
        return (
            <div>
                <h2>404</h2> 
                <h4>Could not find the page <code>{window.location.pathname}</code></h4>
            </div>
        )
    }
}

export default HomePage;