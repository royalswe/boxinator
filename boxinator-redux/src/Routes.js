import React from 'react';
import { Link, Route, Switch } from 'react-router-dom';
import ListItemsPage from './components/views/ListItemsPage';
import AddBoxPage from './components/views/AddBoxPage';
import HomePage from './components/views/HomePage';
import NotFoundPage from './components/views/NotFoundPage';


const Routes = () => (
    <div>
        <p className="App-intro">
        <Link to="/">Home</Link>
        {' | '}
        <Link to="listboxes">Box list</Link>
        {' | '}
        <Link to="addbox">Add new Box</Link>
        </p>
        <Switch>
            <Route exact path="/" component={HomePage} />
            <Route exact path="/listboxes" component={ListItemsPage} />
            <Route exact path="/addbox" component={AddBoxPage} />
            <Route path="*" component={NotFoundPage} />
        </Switch>
    </div>
)

export default Routes;