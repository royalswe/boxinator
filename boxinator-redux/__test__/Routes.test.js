import React from 'react';
import { shallow, mount, render } from 'enzyme';
import { MemoryRouter, Link, Route } from 'react-router-dom';
import { StaticRouter } from 'react-router'
import renderer from 'react-test-renderer';
import Routes from '../src/Routes';
import toJson from 'enzyme-to-json';
import NotFoundPage from '../src/components/views/NotFoundPage';
import AddBoxPage from '../src/components/views/AddBoxPage';
import HomePage from '../src/components/views/HomePage';
import { Provider } from "react-redux";

/**
* middleware function. This is a mockup for the redux-thunk.
*/
const thunk = ({ dispatch, getState }) => next => action => {
if (typeof action === 'function') {
    return action(dispatch, getState)
}
    return next(action)
}
/**
 * fake the redux methods getState, dispatch, and next
 */
const create = () => {
const store = {
    getState: jest.fn(() => ({})),
    dispatch: jest.fn()
};
    const next = jest.fn()
    const invoke = (action) => thunk(store)(next)(action)
    return {store, next, invoke}
};


test('should render the navigation links', () => {
    const wrapper = shallow(<Routes />); 
    const tree = toJson(wrapper);
    expect(tree).toMatchSnapshot();
});

test('should render the HomePage component when visiting start page', () => {
    const component = mount(
        <MemoryRouter initialEntries={['/']} initialIndex={0} >
        <Routes />
        </MemoryRouter>
    )
    expect(component.find(HomePage).length).toBe(1);
});

test('should render the NotFoundPage component when visiting non existing page', () => {
    const component = mount(
        <MemoryRouter initialEntries={['/dont-exist']} initialIndex={0} >
        <Routes />
        </MemoryRouter>
    )
    expect(component.find(NotFoundPage).length).toBe(1);
    expect(component.find(HomePage).length).toBe(0);
});


test('should render the AddBoxPage component when visiting /addbox', () => {
    /**
     * create fake store to be able to generate page
     */
    const { store, invoke } = create()
    invoke((dispatch, getState) => {
      dispatch('LOAD_ITEM_SUCCESS')
      getState();
    })

    const component = mount(
        <Provider store={store}>
            <MemoryRouter initialEntries={['/addbox']} initialIndex={0} >
                <Routes />
            </MemoryRouter>
        </Provider>
    )

    expect(component.find(NotFoundPage).length).toBe(0);
    expect(component.find(HomePage).length).toBe(0);
    expect(component.find(AddBoxPage).length).toBe(1);
});
