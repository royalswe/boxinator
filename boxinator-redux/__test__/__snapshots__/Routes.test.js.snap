// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render the navigation links 1`] = `
<div>
  <p
    className="App-intro"
  >
    <Link
      replace={false}
      to="/"
    >
      Home
    </Link>
     | 
    <Link
      replace={false}
      to="listboxes"
    >
      Box list
    </Link>
     | 
    <Link
      replace={false}
      to="addbox"
    >
      Add new Box
    </Link>
  </p>
  <Switch>
    <Route
      component={[Function]}
      exact={true}
      path="/"
    />
    <Route
      component={[Function]}
      exact={true}
      path="/listboxes"
    />
    <Route
      component={[Function]}
      exact={true}
      path="/addbox"
    />
    <Route
      component={[Function]}
      path="*"
    />
  </Switch>
</div>
`;
